### PR TITLE
export unstructured helper function nestedFieldNoCopy

### PR DESCRIPTION
**Pull request by skriss**
Signed-off-by: Steve Kriss <steve@heptio.com>

**What this PR does / why we need it**: Export the unstructured helper function `nestedFieldNoCopy`.  This enables checking for existence of nested fields without requiring a deep-copy via JSON. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by ncdc** [comment]
/ok-to-test
@kubernetes/sig-api-machinery-pr-reviews

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 1]
```diff
@@ -149,6 +149,21 @@ func NestedSlice(obj map[string]interface{}, fields ...string) ([]interface{}, b
 	return runtime.DeepCopyJSONValue(val).([]interface{}), true, nil
 }
 
+// NestedSliceExists checks for existence of a nested field of []interface{}.
+// Returns (true, nil) if a field with the right type exists, (false, nil) if
+// it does not exist, or (false, error) if an error is encountered traversing
+// obj.
+func NestedSliceExists(obj map[string]interface{}, fields ...string) (bool, error) {
+	val, found, err := nestedFieldNoCopy(obj, fields...)
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: Why not just expose this function and let people do what they need to do.  These helpers seem overly restrictive for how you can use a thing.

**Comment by sttts** [comment]
Lgtm. But would like to hear @ash2k's opinion. Holding until then.

/hold

**Comment by ash2k** [comment]
I think we can merge this as is but I agree with @deads2k - we could expose the underlying helper functions. Maybe do both things - add these two helpers and make those private functions public?

**Comment by deads2k** [comment]
> I think we can merge this as is but I agree with @deads2k - we could expose the underlying helper functions. Maybe do both things - add these two helpers and make those private functions public?

Unused helpers always seem odd.  Having been responsible for a few exposed functions that *could* be private, I'm fine with exposing `nestedFieldNoCopy` to allow this use-case to be built externally, but to my knowledge we don't have any `*Exists` functions at all.  We can expose a useful function externally, limit our surface area to what we need, and the external integration can still be built.

I'm assuming it's a external thing or there would be a change to use these functions somewhere in the code-base, right?

**Comment by sttts** [comment]
> we don't have any *Exists functions at all.

Because the non-object getters give you the `exists` bool as well. For slices and maps we distinguish between "copy" and "no-copy".

**Comment by skriss** [comment]
yes, this is for an external integration. I'd be okay with exposing `nestedFieldNoCopy` if that's what everyone prefers since that addresses the main limitation we're trying to address (can't tell if a nested field exists without a deep copy).

**Comment by deads2k** [comment]
> yes, this is for an external integration. I'd be okay with exposing nestedFieldNoCopy if that's what everyone prefers since that addresses the main limitation we're trying to address (can't tell if a nested field exists without a deep copy).

Let's do that.

**Comment by ncdc** [comment]
@sttts @ash2k @deads2k better? Ok to remove the hold?

**Comment by wenjiaswe** [comment]
cc @roycaihw

**Pull request merged**